File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<int, vector<int>, greater<int>> pq;

        for(int i=0; i<lists.size(); i++){
            ListNode* mover = lists[i];
            while(mover){
                pq.push(mover->val);
                mover = mover->next;
            }
        }

        ListNode* head = new ListNode();
        ListNode* mover = head;

        while(!pq.empty()){
            mover->next = new ListNode(pq.top());
            pq.pop();

            mover = mover->next;
        }

        return head->next;
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
   let dummy = new ListNode();
   let mover = dummy;

    while(true){
        let minIndex = -1;
        for(let i=0; i<lists.length ; i++){
            if(lists[i] && (minIndex == -1 || (lists[i].val < lists[minIndex].val))){
                minIndex = i;
            }
        }

        if(minIndex == -1) break;

        mover.next = lists[minIndex];
        mover = mover.next;
        lists[minIndex] = lists[minIndex].next;
    }

    return dummy.next;
};